File: lint-arwaky/src-python/taxonomy/source_parser_error.py
"""source_parser_error — Source parsing domain error types."""

from pydantic import BaseModel, ConfigDict

from .error_code_vo import ErrorCode
from .error_value_vo import (
    ErrorMessage,
    Cause,
)
from .lint_position_vo import (
    LineNumber,
    ColumnNumber,
)
from .file_path_vo import FilePath


class SourceParserError(BaseModel):
    """General failure during source code parsing."""

    model_config = ConfigDict(frozen=True)

    path: FilePath
    message: ErrorMessage
    error_code: ErrorCode | None = None
    cause: Cause | None = None

    def __str__(self):
        code = f" [{self.error_code}]" if self.error_code else ""
        return f"Parser Error on {self.path}{code}: {self.message}"


class SyntaxErrorVO(SourceParserError):
    """Source code contains invalid syntax."""

    line: LineNumber | None = None
    column: ColumnNumber | None = None

    def __str__(self):
        pos = f" at {self.line}:{self.column}" if self.line else ""
        return f"Syntax Error on {self.path}{pos}: {self.message}"

File: lint-arwaky/src-rust/taxonomy/source_parser_error.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SourceParserError {
    pub path: FilePath,
    pub message: ErrorMessage,
    #[serde(default)]
    pub error_code: Option<ErrorCode>,
    #[serde(default)]
    pub cause: Option<Cause>,
}

impl SourceParserError {
    pub fn new(path: FilePath, message: ErrorMessage) -> Self {
        Self { path, message, error_code: None, cause: None }
    }
}

impl std::fmt::Display for SourceParserError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let code = self.error_code.as_ref().map(|c| format!(" [{}]", c)).unwrap_or_default();
        write!(f, "Parser Error on {}{}: {}", self.path, code, self.message)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SyntaxErrorVO {
    #[serde(flatten)]
    pub base: SourceParserError,
    #[serde(default)]
    pub line: Option<LineNumber>,
    #[serde(default)]
    pub column: Option<ColumnNumber>,
}

impl SyntaxErrorVO {
    pub fn new(path: FilePath, message: ErrorMessage) -> Self {
        Self { base: SourceParserError::new(path, message), line: None, column: None }
    }
}

impl std::fmt::Display for SyntaxErrorVO {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let pos = match (&self.line, &self.column) {
            (Some(l), Some(c)) => format!(" at {}:{}", l, c),
            (Some(l), None) => format!(" at {}", l),
            _ => String::new(),
        };
        write!(f, "Syntax Error on {}{}: {}", self.base.path, pos, self.base.message)
    }
}
